umbraco/Umbraco-CMS
# User forgot password functionality for Management API

### Summary
This PR adds three new endpoints
- `/umbraco/management/api/v1/security/forgot-password`
  - Used to request an email with a token.
- `/umbraco/management/api/v1/security/forgot-password/verify`
   - Used to verify the token without resetting. Use case is when you enter the page from the link in a mail, we can say it is used.
- `/umbraco/management/api/v1/security/forgot-password/reset`
   - Used to actually reset the password using the token.

- Also a new filter is introduced to ensure a request has a minimum response time.
- A few small changes in names have been made to align things
- 
### Out of scope
- The generated email + link is out of scope of this PR
   
### Tests
- Use the endpoint to request an email (`/umbraco/management/api/v1/security/forgot-password`)
 - This shown always response slowly so you cannot guess if the user email exists.
  - Remember to setup a fake SMTP and get the link from the email
- Read the email and copy the token from the link
  - A link could look like this: `https://localhost:44339/umbraco/ValidatePasswordResetCode?invite=1e70f841-c261-413b-abb2-2d68cdb96094%257CQ2ZESjhBSGFMTHpmOE45TnNCam9MN2tEK1Q0UjdNZnpjQktnME9RZ3kvQUhtWko0VWJQMU56ODdrWEk1VDl2MldzK1NDT3VZZ05na2VLcmczMVhtdkhob0hwL044WDhwZ0dNRnFQVUdYeWsvN1VhYU90OGlrR0pKbUg5SlA5ZWFqU1R2S3VtU29SRVJSVnJ0NGwwYmNwaUdkWHEwWUZMWHBKblQ1aGpiakxsU28vOFo0dnJnazlwV3NKWkU3MHZKMTRDRkdBPT0`
 The url query is today a combination of the userkey, a seperator and a token. In this case:
      - UserKey: `1e70f841-c261-413b-abb2-2d68cdb96094` 
      - Seperator: `%257C` 
      - Token: `Q2ZESjhBSGFMTHpmOE45TnNCam9MN2tEK1Q0UjdNZnpjQktnME9RZ3kvQUhtWko0VWJQMU56ODdrWEk1VDl2MldzK1NDT3VZZ05na2VLcmczMVhtdkhob0hwL044WDhwZ0dNRnFQVUdYeWsvN1VhYU90OGlrR0pKbUg5SlA5ZWFqU1R2S3VtU29SRVJSVnJ0NGwwYmNwaUdkWHEwWUZMWHBKblQ1aGpiakxsU28vOFo0dnJnazlwV3NKWkU3MHZKMTRDRkdBPT0`
 
 - Use the `/umbraco/management/api/v1/security/forgot-password/verify` endpoint to verify the token is correct.
   - This shown always response slowly so you cannot guess if the user key exists.
   - Also test unhappy paths, with invalid token and user key. 
  - Use the `/umbraco/management/api/v1/security/forgot-password/reset` endpoint.
    - This shown always response slowly so you cannot guess if the user key exists.
    - First test unhappy paths
       - Invalid user key
       - Invalid token
       - Invalid password combination (e.g. too short)
     - Then test happy path.
 - Verify you can sign-in with the newly updated password
 - Try to change the minimum response time value to something else then the default
   - `Umbraco:CMS:Security:UserPassword:MinimumResponseTime` = "0.00.00.00.001"
